pkg/addonmanager/controllers/agentdeploy/hosting_controller.go

```diff
@@ -0,0 +1,346 @@
+	"open-cluster-management.io/addon-framework/pkg/basecontroller/factory"
+	"open-cluster-management.io/addon-framework/pkg/utils"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	addonv1alpha1client "open-cluster-management.io/api/client/addon/clientset/versioned"
+	addoninformerv1alpha1 "open-cluster-management.io/api/client/addon/informers/externalversions/addon/v1alpha1"
+	addonlisterv1alpha1 "open-cluster-management.io/api/client/addon/listers/addon/v1alpha1"
+	clusterinformers "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterlister "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	workv1client "open-cluster-management.io/api/client/work/clientset/versioned"
+	workinformers "open-cluster-management.io/api/client/work/informers/externalversions/work/v1"
+	worklister "open-cluster-management.io/api/client/work/listers/work/v1"
+	workapiv1 "open-cluster-management.io/api/work/v1"
+)
+
+// addonHostingDeployController deploy addon agent resources on the hosting cluster in Hosted mode.
+type addonHostingDeployController struct {
+	workClient                workv1client.Interface
+	addonClient               addonv1alpha1client.Interface
+	managedClusterLister      clusterlister.ManagedClusterLister
+	managedClusterAddonLister addonlisterv1alpha1.ManagedClusterAddOnLister
+	workLister                worklister.ManifestWorkLister
+	agentAddons               map[string]agent.AgentAddon
+	cache                     *workCache
+}
+
+func NewAddonHostingDeployController(
+	workClient workv1client.Interface,
+	addonClient addonv1alpha1client.Interface,
+	clusterInformers clusterinformers.ManagedClusterInformer,
+	addonInformers addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	workInformers workinformers.ManifestWorkInformer,
+	agentAddons map[string]agent.AgentAddon,
+) factory.Controller {
+	c := &addonHostingDeployController{
+		workClient:                workClient,
+		addonClient:               addonClient,
+		managedClusterLister:      clusterInformers.Lister(),
+		managedClusterAddonLister: addonInformers.Lister(),
+		workLister:                workInformers.Lister(),
+		agentAddons:               agentAddons,
+		cache:                     newWorkCache(),
+	}
+
+	return factory.New().WithFilteredEventsInformersQueueKeysFunc(
+		func(obj runtime.Object) []string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return []string{key}
+		},
+		func(obj interface{}) bool {
+			accessor, _ := meta.Accessor(obj)
+			if _, ok := c.agentAddons[accessor.GetName()]; !ok {
+				return false
+			}
+
+			return true
+		},
+		addonInformers.Informer()).
+		WithFilteredEventsInformersQueueKeysFunc(
+			func(obj runtime.Object) []string {
+				accessor, _ := meta.Accessor(obj)
+				return []string{
+					fmt.Sprintf("%s/%s",
+						accessor.GetLabels()[constants.AddonNamespaceLabel],
+						accessor.GetLabels()[constants.AddonLabel],
```

Comment:
the original controller should filter out these works.

Reply:
yes, they are filtered by `if accessor.GetName() != constants.DeployWorkName(addonName) {` already.